Signing TAs using subkeys

jenswi-linaro opened the pull request:
This implements subkeys as a way of supporting multiple independent signers for TAs.

Replaces the RFC https://github.com/OP-TEE/optee_os/pull/5523

I've addressed comments from https://github.com/OP-TEE/optee_os/pull/5523 and https://github.com/OP-TEE/optee_test/pull/615

Documentation at https://github.com/OP-TEE/optee_docs/pull/170

I've also refactored the sign_encrypt.py script to make it easier to understand and maintain.
We're using both logger.error() + sys.exit() and raise Exception() to report various errors. The purpose of logger.error()/logger.info() isn't entirely clear to me. I wouldn't mind using just print() instead of logger. Perhaps something to address in another PR.

jenswi-linaro commented:
There are a few checkpatch warnings that I'm not sure it makes sense to try to fix.

jforissier commented on scripts/sign_encrypt.py:
```diff
@@ -29,391 +59,876 @@ def int_parse(str):
+                if enc_algo not in enc_tee_alg.values():
+                    raise Exception('Unrecognized encrypt algorithm: 0x{:08x}'
+                                    .format(enc_value))
+                print('  nonce_len:  {} (bytes)'.format(nonce_len))
+                if nonce_len != NONCE_SIZE:
+                    raise Exception("Unexpected nonce len: {}"
+                                    .format(nonce_len))
+                nonce = self.inf[offs:offs + nonce_len]
+                print('  nonce:      {}...'.format(binascii.hexlify(nonce)))
+                offs += nonce_len
+
+                print('  tag_len:    {} bytes'.format(tag_len))
```
It would be better to use the exact same name as in the C structure: `iv_size`, `tag_size`. Maybe print/decode flags too while we're at it?

jforissier commented on scripts/sign_encrypt.py:
```diff
@@ -29,391 +59,876 @@ def int_parse(str):
+                tag = self.inf[-tag_len:]
+                print('  tag:        {}...'.format(binascii.hexlify(tag)))
+                ciphertext = self.inf[offs:-tag_len]
+                print('  Ciph offset:{} (0x{:x}) bytes)'.format(offs, offs))
+                print('  Ciph size:  {} (0x{:x}) (bytes)'
+                      .format(len(ciphertext), len(ciphertext)))
+                if len(ciphertext) != img_size:
+                    raise Exception("Unexpected ciphertext size: ",
+                                    "got {}, expected {}"
+                                    .format(len(ciphertext), img_size))
```
Would it make sense not to make a special case of this and print "TA offset" and "TA size" like in the else part? It is ciphertext, but it is also the TA;  the fact it is encrypted doesn't matter much and it is shown by the `SHDR_ENCRYPTED_TA` type anyways.

jforissier commented on scripts/sign_encrypt.py:
```diff
@@ -29,391 +59,876 @@ def int_parse(str):
+                      .format(offs, offs))
+
+            shdr = self.inf[offs:offs + SHDR_SIZE]
+            [magic, img_type, img_size, algo_value, digest_len,
+             sig_len] = struct.unpack('<IIIIHH', shdr)
```
`s/digest_len/hash_size/` and `s/sig_len/sig_size/` to match the C definition and the text below

jforissier reviewed (commented):
> There are a few checkpatch warnings that I'm not sure it makes sense to try to fix.

The one about TEE_UUID we can leave, checkpatch is clearly confused and the code is fine.
As for the other one see my suggestion below.
I have not done a full review yet, I will focus in the doc PR first.

jforissier commented:
For "sign_encrypt.py: make logger global":
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier commented:
For "sign_encrypt.py: refactor argument parsing":
That's a very welcome refactoring, things were starting to get messy!

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier reviewed (commented):
For commit "sign_encrypt.py: refactor with Binary_image class":
One comment below, then:
`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier commented on scripts/sign_encrypt.py:
```diff
@@ -222,7 +227,6 @@ def assign_default_value(parsed, attr, func):
 
     return parsed
 
-
```
That will make `pycodestyle` unhappy. CI hasn't detected this and many other coding style issues due to the checkpatch errors which make the CI script exit without even running `pycodestyle`. I need to see how to fix this. In the meantime, could you run `pycodestyle` on each commit and make sure any issues are addressed?

jforissier reviewed (commented):
For commit "sign_encrypt.py: add command func in argument parsing":
With the Python coding style issues addressed:
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier commented on scripts/sign_encrypt.py:
```diff
@@ -381,6 +390,109 @@ def parse(self):
+                print('  nonce:      {}...'.format(binascii.hexlify(nonce)))
+                offs += nonce_len
+
+                print('  tag_len:    {} bytes'.format(tag_len))
```
It would be better to use the exact same name as in the C structure: iv_size, tag_size. Maybe print/decode flags too while we're at it?

jforissier commented on scripts/sign_encrypt.py:
```diff
@@ -381,6 +390,109 @@ def parse(self):
+                      .format(len(img), len(img)))
+                if len(img) != img_size:
+                    raise Exception("Unexpected img size: got {}, expected {}"
+                                    .format(len(img), img_size))
```
Would it make sense not to make a special case of this and print "TA offset" and "TA size" like in the else part? It is ciphertext, but it is also the TA; the fact it is encrypted doesn't matter much and it is shown by the `SHDR_ENCRYPTED_TA` type anyways.

jforissier reviewed (commented):
Comments on "sign_encrypt.py: add display command".

jforissier commented on scripts/sign_encrypt.py:
```diff
@@ -488,10 +606,82 @@ def display(self):
+
+            shdr = self.inf[offs:offs + SHDR_SIZE]
+            [magic, img_type, img_size, algo_value, digest_len,
+             sig_len] = struct.unpack('<IIIIHH', shdr)
```
s/digest_len/hash_size/ and s/sig_len/sig_size/ to match the C definition and the text below

jenswi-linaro replied:
OK, I'll do that in a new commit on top of this since it touches quite a lot.

jforissier commented on scripts/sign_encrypt.py:
```diff
@@ -488,10 +606,82 @@ def display(self):
+
+            img_digest = self.inf[offs:offs + digest_len]
+            print('  hash:       {}...'
+                  .format(binascii.hexlify(img_digest[:12])))
```
Why truncate?

jforissier replied:
SUre but honestly having a decoding tool print slightly long data isn't unexpected, and 80 character is not that much. It might come in handy to troubleshoot corruption some day.
Then what about showing the start of `struct shdr` and`struct shdr_subkey`? It would help understand the fact that `algo` is present twice in the same block of data. BTW the algo is not decoded to a name the second time.
Also, drop the Python binary string prefix (b'xxx').
```diff
diff --git a/scripts/sign_encrypt.py b/scripts/sign_encrypt.py
index 1141920af..924c2f664 100755
--- a/scripts/sign_encrypt.py
+++ b/scripts/sign_encrypt.py
@@ -556,6 +556,7 @@ class Binary_image:
 
         def display_ta():
             nonlocal offs
+            print(' struct shdr_bootstrap_ta')
             ta_uuid = self.inf[offs:offs + UUID_SIZE]
             print('  uuid:       {}'.format(uuid.UUID(bytes=ta_uuid)))
             offs += UUID_SIZE
@@ -569,6 +570,7 @@ class Binary_image:
                 [enc_algo, flags, nonce_len,
                  tag_len] = struct.unpack('<IIHH', ehdr)
 
+                print(' struct shdr_encrypted_ta')
                 enc_algo_name = 'Unkown'
                 if enc_algo in enc_tee_alg.values():
                     enc_algo_name = value_to_key(enc_tee_algo, enc_algo)
@@ -600,8 +602,8 @@ class Binary_image:
                                     .format(len(ciphertext), img_size))
             else:
                 img = self.inf[offs:]
-                print('  TA offset:  {} (0x{:x}) bytes'.format(offs, offs))
-                print('  TA size:    {} (0x{:x}) bytes'
+                print(' TA offset:  {} (0x{:x}) bytes'.format(offs, offs))
+                print(' TA size:    {} (0x{:x}) bytes'
                       .format(len(img), len(img)))
                 if len(img) != img_size:
                     raise Exception("Unexpected img size: got {}, expected {}"
@@ -642,6 +644,7 @@ class Binary_image:
             if algo_value in sig_tee_alg.values():
                 algo_name = value_to_key(sig_tee_alg, algo_value)
 
+            print(' struct shdr')
             print('  magic:      0x{:08x}'.format(magic))
             print('  img_type:   {} ({})'.format(img_type, img_type_name))
             print('  img_size:   {} bytes'.format(img_size))
@@ -658,8 +661,8 @@ class Binary_image:
                                     .format(digest_len))
 
             img_digest = self.inf[offs:offs + digest_len]
-            print('  hash:       {}...'
-                  .format(binascii.hexlify(img_digest[:12])))
+            print('  hash:       {}'
+                  .format(binascii.hexlify(img_digest).decode('ascii')))
             offs += digest_len
             sig = self.inf[offs:offs + sig_len]
             offs += sig_len
@@ -672,6 +675,7 @@ class Binary_image:
                                       offs + UUID_SIZE + SK_HDR_SIZE]
                 [name_size, subkey_version, max_depth, algo,
                  attr_count] = struct.unpack('<IIIII', img_subkey)
+                print(' struct shdr_subkey')
                 print('  uuid:       {}'.format(uuid.UUID(bytes=img_uuid)))
                 print('  name_size:  {}'.format(name_size))
                 print('  subkey_version: {}'.format(subkey_version))
```
```
Subkey
 struct shdr
  magic:      0x4f545348
  img_type:   3 (SHDR_SUBKEY)
  img_size:   320 bytes
  algo:       0x70414930 (TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA256)
  hash_size:  32 bytes
  sig_size:   256 bytes
  hash:       0386b87b5832e571752bd96c9db8cfade9807db2b926f8529fcd1953362cc5d2
 struct shdr_subkey
  uuid:       fd02c9da-306c-48c7-a49c-bbd827ae86ee
  name_size:  0
  subkey_version: 0
  max_depth:  2
  algo:       0x70414930
  attr_count: 2
  next name:  ""
Next header at offset: 628 (0x274)
Subkey
 struct shdr
  magic:      0x4f545348
  img_type:   3 (SHDR_SUBKEY)
  img_size:   320 bytes
  algo:       0x70414930 (TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA256)
  hash_size:  32 bytes
  sig_size:   256 bytes
  hash:       c774cc6883367487e5a94c2263fab556984c09d30ee1a7cb95f1ac8f4940eed9
 struct shdr_subkey
  uuid:       fd02c9da-306c-48c7-a49c-bbd827ae86ee
  name_size:  0
  subkey_version: 0
  max_depth:  1
  algo:       0x70414930
  attr_count: 2
  next name:  ""
Next header at offset: 1256 (0x4e8)
Bootstrap TA
 struct shdr
  magic:      0x4f545348
  img_type:   1 (SHDR_BOOTSTRAP_TA)
  img_size:   222064 bytes
  algo:       0x70414930 (TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA256)
  hash_size:  32 bytes
  sig_size:   256 bytes
  hash:       409905657c6168257ba85d98f0e2529e1672c34954345416cf43a9a0d04051ca
 struct shdr_bootstrap_ta
  uuid:       fd02c9da-306c-48c7-a49c-bbd827ae86ee
  ta_version: 0
 TA offset:  1584 (0x630) bytes
 TA size:    222064 (0x36370) bytes
```

jforissier reviewed (commented):
Comments on "sign_encrypt.py: add subkey support"

jforissier reviewed (commented):
For commit "core: ree_fs: refactor check_update_version()":
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

etienne-lms commented on core/crypto/signed_hdr.c:
```diff
@@ -87,3 +90,266 @@ TEE_Result shdr_verify_signature(const struct shdr *shdr)
+		uint8_t digest[TEE_SHA1_HASH_SIZE];
+		TEE_UUID uuid;
+		char name_str[];
+	} *tmp;
```
`= NULL`

jforissier commented:
Comment on "sign_encrypt.py: add subkey support".

`--in` and `--out` are re-used for subkeys, so either the description should be changed or different parameter should be used (`--inskey` / `--outskey` maybe?).

jforissier commented on core/crypto/signed_hdr.c:
```diff
@@ -87,3 +90,266 @@ TEE_Result shdr_verify_signature(const struct shdr *shdr)
+
+	key = calloc(1, sizeof(*key));
+	if (!key)
+		return TEE_ERROR_SECURITY;
```
`TEE_ERROR_OUT_OF_MEMORY` would be more helpful, any concern with it?

jenswi-linaro replied:
No, I'll change.

jforissier commented on core/crypto/signed_hdr.c:
```diff
@@ -87,3 +90,266 @@ TEE_Result shdr_verify_signature(const struct shdr *shdr)
+		goto out_ctx;
+
+	if (subkey->max_depth >= max_depth) {
+		res = TEE_ERROR_SECURITY;
```
`DMSG()`?

jforissier commented on core/crypto/signed_hdr.c:
```diff
@@ -87,3 +90,266 @@ TEE_Result shdr_verify_signature(const struct shdr *shdr)
+		res = TEE_ERROR_SECURITY;
+		goto out_ctx;
+	}
+	if (next_uuid && memcmp(next_uuid, subkey->uuid, sizeof(TEE_UUID))) {
```
`DMSG()`?

jforissier commented on core/kernel/ree_fs_ta.c:
```diff
@@ -164,12 +267,89 @@ static TEE_Result ree_fs_ta_open(const TEE_UUID *uuid,
+	 * If we're verifying this TA using a subkey, make sure that
+	 * the UUID of the TA belongs to the namespace defined by the subkey.
+	 * The namespace is defined as in RFC4122, that is, valid UUID
+	 * is calculated as a V5 UUID SHA-1(subkey UUID, "name string").
```
SHA-512

jforissier commented:
My last round of comments for today, about small issues I noticed with `sign_encrypt.py`:

- `--subkey-version` should have a default value
- `sign-subkey` without `--max-depth` causes an exception
-  `sign-enc` should probably reject multiple `--subkey` parameters since it does not concatenate them. For other parameters it is clear that subsequent use of the same will override the value, but for subkeys it is confusing.
- `subkey-uuid` without parameters causes an exception
- `subkey-uuid --in subkey.bin --name xyz` generates a new UUID even when the subkey `name_size` is zero. My understanding is that such a subkey is an identity subkey which mandates the use of the same UUID for the next subkey or TA.

jenswi-linaro commented:
Tags applied and pycodestyle issues addressed. Force push since the deltas with the style issues wouldn't be enjoyable for anyone. I'll continue with the rest of the review comments now.

jforissier commented on scripts/sign_encrypt.py:
```diff
@@ -29,391 +61,845 @@ def int_parse(str):
+            else:
+                max_depth = 0
+        if max_depth < 0:
+            raise Exception("Cannot use max_depth {}".format(max_depth))
```
General comment: since we're using Python 3, f-strings could help make things more compact (`f"Cannot use max_depth {max_depth}"`).

jenswi-linaro replied:
OK if we take this in a new commit on top?

jforissier replied:
Whatever is easier for you. I don't mind reviewing the whole series again if you prefer.

jenswi-linaro replied:
I read that f-strings are a python 3.6 thing and we actually don't use f-strings in any python-script in this git, yet. I'd rather have that concept introduced in a separate PR i you don't mind.

jforissier replied:
> we actually don't use f-strings in any python-script in this git, yet

Right, in fact I think I am the one who introduced them in the (now deleted) file `dump_ta_header.py`. Let's deal with that later especially since it requires Python 3.6+.

jenswi-linaro commented:
> Comment on "sign_encrypt.py: add subkey support".
>
>--in and --out are re-used for subkeys, so either the description should be changed or different parameter should be used (--inskey / --outskey maybe?).

Updated description for --in and --out.

> My last round of comments for today, about small issues I noticed with `sign_encrypt.py`:
> 
> * `--subkey-version` should have a default value

Fixed
> * `sign-subkey` without `--max-depth` causes an exception

You may have been testing with something old, couldn't reproduce
> * `sign-enc` should probably reject multiple `--subkey` parameters since it does not concatenate them. For other parameters it is clear that subsequent use of the same will override the value, but for subkeys it is confusing.

Fixed
> * `subkey-uuid` without parameters causes an exception

Fixed
> * `subkey-uuid --in subkey.bin --name xyz` generates a new UUID even when the subkey `name_size` is zero. My understanding is that such a subkey is an identity subkey which mandates the use of the same UUID for the next subkey or TA.

Fixed, I added a check to see that the name wasn't larger than the max name size.

jforissier commented:
> > * `sign-subkey` without `--max-depth` causes an exception
> 
> You may have been testing with something old, couldn't reproduce

Here is a reproducer, you need 2 subkeys apparently.
```
openssl genrsa -out keys/subkey1.pem 2048
openssl genrsa -out keys/subkey2.pem 2048
scripts/sign_encrypt.py sign-subkey --in keys/subkey1.pem --uuid fd02c9da-306c-48c7-a49c-bbd827ae86ee --key keys/default_ta.pem --out out/subkey1.bin --name-size 0 --subkey-version 0
scripts/sign_encrypt.py sign-subkey --in keys/subkey2.pem --uuid fd02c9da-306c-48c7-a49c-bbd827ae86ee --key keys/subkey1.pem --out out/subkey2.bin --subkey out/subkey1.bin --name-size 0 --subkey-version 0
```

jenswi-linaro commented:
> > > * `sign-subkey` without `--max-depth` causes an exception
> > 
> > 
> > You may have been testing with something old, couldn't reproduce
> 
> Here is a reproducer, you need 2 subkeys apparently.
> 
> ```
> openssl genrsa -out keys/subkey1.pem 2048
> openssl genrsa -out keys/subkey2.pem 2048
> scripts/sign_encrypt.py sign-subkey --in keys/subkey1.pem --uuid fd02c9da-306c-48c7-a49c-bbd827ae86ee --key keys/default_ta.pem --out out/subkey1.bin --name-size 0 --subkey-version 0
> scripts/sign_encrypt.py sign-subkey --in keys/subkey2.pem --uuid fd02c9da-306c-48c7-a49c-bbd827ae86ee --key keys/subkey1.pem --out out/subkey2.bin --subkey out/subkey1.bin --name-size 0 --subkey-version 0
> ```

`out/subkey1.bin` has max depth 0 so default value for subkey2 becomes -1. I'll try to improve error message and add a check that the max depth can be used.

jenswi-linaro commented:
Squashed in fixups and addressed the comments on the display sub-command.

jenswi-linaro commented on core/crypto/signed_hdr.c:
```diff
@@ -87,3 +90,266 @@ TEE_Result shdr_verify_signature(const struct shdr *shdr)
+		res = load_rsa_key(subkey, &key->pub_key.rsa);
+		break;
+	default:
+		panic();
```
Well spotted, I'll return `TEE_ERROR_SECURITY` instead.

jenswi-linaro commented on core/kernel/ree_fs_ta.c:
```diff
@@ -67,13 +67,118 @@ struct ree_fs_ta_handle {
+			res = TEE_ERROR_ACCESS_CONFLICT;
+			goto out;
+		} else if (db_entry.version < version) {
+			memcpy(db_entry.uuid, uuid, sizeof(TEE_UUID));
```
I'll remove it.

jenswi-linaro commented on ta/arch/arm/link.mk:
```diff
@@ -108,10 +108,11 @@ cmd-echo$(user-ta-uuid) := SIGNENC
 $(link-out-dir$(sm))/$(user-ta-uuid).ta: \
 			$(link-out-dir$(sm))/$(user-ta-uuid).stripped.elf \
-			$(TA_SIGN_KEY) \
+			$(TA_SIGN_KEY) $(TA_SUBKEY_DEPS) \
```
I'll add something to mk/config.mk

jforissier commented:
In commit "sign_encrypt.py: refactor with Binary_image class": you forgot to s/Binary_image/BinaryImage/ in the commit description.

jenswi-linaro commented:
Addressing comments on "core: support loading TAs signed with a subkey".

Regarding the DMSG()s, I'm not so keen on adding debug prints in the TA verification. At least not in the PR. Perhaps better to add a new PR to propose something if needed.

jforissier commented on scripts/sign_encrypt.py:
```diff
@@ -371,6 +380,116 @@ def parse(self):
+                    raise Exception("Unexpected nonce len: {}"
+                                    .format(nonce_len))
+                nonce = self.inf[offs:offs + nonce_len]
+                print('  iv:         {}...'.format(binascii.hexlify(nonce)))
```
```
        print('  iv:         {}'
              .format(binascii.hexlify(nonce).decode('ascii')))
```

jforissier commented on scripts/sign_encrypt.py:
```diff
@@ -371,6 +380,116 @@ def parse(self):
+                enc_algo_name = 'Unkown'
+                if enc_algo in enc_tee_alg.values():
+                    enc_algo_name = value_to_key(enc_tee_algo, enc_algo)
+                print('  enc_algo:   {} ({})'.format(enc_algo, enc_algo_name))
```
Use `0x{:08x}` for the algo value

jforissier commented on scripts/sign_encrypt.py:
```diff
@@ -371,6 +380,116 @@ def parse(self):
+                print(' struct shdr_encrypted_ta')
+                enc_algo_name = 'Unkown'
+                if enc_algo in enc_tee_alg.values():
+                    enc_algo_name = value_to_key(enc_tee_algo, enc_algo)
```
s/enc_tee_algo/enc_tee_alg/

jforissier reviewed (commented):
Comments on "sign_encrypt.py: add display command".

There is an issue with the display of encrypted TAs which I have not been able to figure out at first glance. Two examples:

```
scripts/sign_encrypt.py sign-enc --uuid fd02c9da-306c-48c7-a49c-bbd827ae86ee --in out/arm/ta/pkcs11/fd02c9da-306c-48c7-a49c-bbd827ae86ee.stripped.elf --out out/arm/ta/pkcs11/fd02c9da-306c-48c7-a49c-bbd827ae86ee.ta --key keys/default_ta.pem --enc-key 01020304050607080807060504030201

scripts/sign_encrypt.py display --in out/arm/ta/pkcs11/fd02c9da-306c-48c7-a49c-bbd827ae86ee.ta
Encrypted TA
 struct shdr
  magic:      0x4f545348
  img_type:   2 (SHDR_ENCRYPTED_TA)
  img_size:   222064 bytes
  algo:       0x70414930 (TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA256)
  hash_size:  32 bytes
  sig_size:   256 bytes
  hash:       ebb0945405c208d6c9f5321a6b502e6fbd4d0f1dc8d0f3c5da379c323b8c0470
 struct shdr_bootstrap_ta
  uuid:       fd02c9da-306c-48c7-a49c-bbd827ae86ee
  ta_version: 0
 struct shdr_encrypted_ta
  enc_algo:   0x40000810 (TEE_ALG_AES_GCM)
  flags:      0x0 (SHDR_ENC_KEY_DEV_SPECIFIC)
  iv_size:    12 (bytes)
  iv:         b'227c5b84fbb0f0c2defd3d0f'...
  tag_size:   16 bytes
  tag:        b'27becc04842c8aa1bbf7635ad1887610'...
 TA offset:  352 (0x160) bytes
 TA size:    222064 (0x36370) bytes
Traceback (most recent call last):
  File "/home/jerome/work/optee_repo_qemu/optee_os/scripts/sign_encrypt.py", line 963, in <module>
    main()
  File "/home/jerome/work/optee_repo_qemu/optee_os/scripts/sign_encrypt.py", line 959, in main
    args.func(args)
  File "/home/jerome/work/optee_repo_qemu/optee_os/scripts/sign_encrypt.py", line 921, in command_display
    ta_image.display()
  File "/home/jerome/work/optee_repo_qemu/optee_os/scripts/sign_encrypt.py", line 641, in display
    name_img = self.inf[offs:offs + name_size]
UnboundLocalError: local variable 'name_size' referenced before assignment

```
```
openssl genrsa -out keys/subkey1.pem 2048

scripts/sign_encrypt.py sign-subkey --in keys/subkey1.pem --uuid fd02c9da-306c-48c7-a49c-bbd827ae86ee --key keys/default_ta.pem --out out/subkey1.bin --name-size 0 --max-depth 0

scripts/sign_encrypt.py sign-enc --uuid fd02c9da-306c-48c7-a49c-bbd827ae86ee --in out/arm/ta/pkcs11/fd02c9da-306c-48c7-a49c-bbd827ae86ee.stripped.elf --subkey out/subkey1.bin --out out/arm/ta/pkcs11/fd02c9da-306c-48c7-a49c-bbd827ae86ee.ta --key keys/default_ta.pem --enc-key 01020304050607080807060504030201

scripts/sign_encrypt.py display --in out/arm/ta/pkcs11/fd02c9da-306c-48c7-a49c-bbd827ae86ee.ta
Subkey
 struct shdr
  magic:      0x4f545348
  img_type:   3 (SHDR_SUBKEY)
  img_size:   320 bytes
  algo:       0x70414930 (TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA256)
  hash_size:  32 bytes
  sig_size:   256 bytes
  hash:       1f0ddc107472b83744567cc77e8cccd638237fb87e9a8177864144c30509636d
 struct shdr_subkey
  uuid:       fd02c9da-306c-48c7-a49c-bbd827ae86ee
  name_size:  0
  subkey_version: 0
  max_depth:  0
  algo:       0x70414930 (TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA256)
  attr_count: 2
  next name:  ""
Next header at offset: 628 (0x274)
Encrypted TA
 struct shdr
  magic:      0x4f545348
  img_type:   2 (SHDR_ENCRYPTED_TA)
  img_size:   222064 bytes
  algo:       0x70414930 (TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA256)
  hash_size:  32 bytes
  sig_size:   256 bytes
  hash:       39a352a6039c733315cb097a145271cb43f49468e8c43600f8cd756097170d0b
 struct shdr_bootstrap_ta
  uuid:       fd02c9da-306c-48c7-a49c-bbd827ae86ee
  ta_version: 0
 struct shdr_encrypted_ta
  enc_algo:   0x40000810 (TEE_ALG_AES_GCM)
  flags:      0x0 (SHDR_ENC_KEY_DEV_SPECIFIC)
  iv_size:    12 (bytes)
  iv:         b'a529c792d178c95d19cfbd87'...
  tag_size:   16 bytes
  tag:        b'88401bd6474bf14e4110cf8a800b63d5'...
 TA offset:  980 (0x3d4) bytes
 TA size:    222064 (0x36370) bytes
  next name:  ""
Next header at offset: 223044 (0x36744)
Traceback (most recent call last):
  File "/home/jerome/work/optee_repo_qemu/optee_os/scripts/sign_encrypt.py", line 963, in <module>
    main()
  File "/home/jerome/work/optee_repo_qemu/optee_os/scripts/sign_encrypt.py", line 959, in main
    args.func(args)
  File "/home/jerome/work/optee_repo_qemu/optee_os/scripts/sign_encrypt.py", line 921, in command_display
    ta_image.display()
  File "/home/jerome/work/optee_repo_qemu/optee_os/scripts/sign_encrypt.py", line 649, in display
    sig_size] = struct.unpack('<IIIIHH', shdr)
struct.error: unpack requires a buffer of 20 bytes
```

jforissier reviewed (commented):
For "sign_encrypt.py: rename to hash_size and sig_size":

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jenswi-linaro commented:
Force push to get "[review] sign_encrypt.py: refactor with Binary_image class" and "[review] sign_encrypt.py: add display command" earlier in the chain.

Updated "sign_encrypt.py: add subkey support" as requested. Tags applied.

jenswi-linaro commented:
The encrypted TA problem should be fixed now.

jforissier commented:
> Addressing comments on "core: support loading TAs signed with a subkey".
> 
> Regarding the DMSG()s, I'm not so keen on adding debug prints in the TA verification. At least not in the PR. Perhaps better to add a new PR to propose something if needed.

That's fine.

jforissier commented:
For "core: support loading TAs signed with a subkey" with fixups, and please do not forget to update the output of `sign_encrypt.py display`:
`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`
`Tested-by: Jerome Forissier <jerome.forissier@linaro.org> (vexpress-qemu_virt)`

jenswi-linaro commented:
Squashed and tags applied. Note that "sign_encrypt.py: add display command" still is missing an A-B or R-B tag.

jforissier commented:
> Squashed and tags applied. Note that "sign_encrypt.py: add display command" still is missing an A-B or R-B tag.

There you go:
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jenswi-linaro commented:
Thanks, tag applied.

jenswi-linaro commented on scripts/sign_encrypt.py:
```diff
@@ -403,7 +407,9 @@ def sign(self):
-        with open(args, 'r') as f:
+        import base64
+
+        with open(sigf, 'r') as f:
```
Fixed

etienne-lms reviewed (commented):
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>`

jenswi-linaro commented:
Tag applied
